### PR TITLE
chore: add flake8 plugins, comprehension, bugbear, builtins

### PR DESCRIPTION
**Pull request by noahnu**
## Description

Adds:
- https://github.com/adamchainz/flake8-comprehensions
- https://github.com/PyCQA/flake8-bugbear
- https://github.com/gforcada/flake8-builtins

## Checklist

- [x] This PR has sufficient test coverage.
- [x] Licenses of new dependencies are compatible with Apache 2.0.

## Additional Comments

The comprehension rule provides a minuscule performance benefit, however it ultimately brings us closer to consistently using "one pattern" when two are available (consistent code).

### CODE REVIEW

File: src/syrupy/serializers/amber.py [hunk 1]
```diff
@@ -163,19 +184,28 @@ def serialize_iterable(
...
     @classmethod
     def serialize(
-        cls, data: "SerializableData", *, depth: int = 0, visited: Set[Any] = set()
+        cls,
+        data: "SerializableData",
+        *,
+        depth: int = 0,
+        visited: Optional[Set[Any]] = None,
     ) -> str:
+        visited = visited if visited is not None else set()
```
Comment (iamogbz) [inline src/syrupy/serializers/amber.py]: Wouldn't this also work?
```suggestion
        visited = visited or set()
```

File: src/syrupy/serializers/amber.py [hunk 2]
```diff
@@ -208,7 +238,7 @@ def file_extension(self) -> str:
         return "ambr"
 
     def discover_snapshots(self, filepath: str) -> Set[str]:
-        return set(name for name in DataSerializer.read_file(filepath).keys())
+        return set(DataSerializer.read_file(filepath).keys())
```
Comment (iamogbz) [inline src/syrupy/serializers/amber.py]: 👍

File: src/syrupy/serializers/amber.py [hunk 3]
```diff
@@ -126,9 +139,13 @@ def serialize_set(
...
     def serialize_dict(
-        cls, data: "SerializableData", *, depth: int = 0, visited: Set[Any] = set()
+        cls,
+        data: "SerializableData",
+        *,
+        depth: int = 0,
+        visited: Optional[Set[Any]] = None,
     ) -> str:
-        kwargs = dict(depth=depth + 1, visited=visited)
+        kwargs = {"depth": depth + 1, "visited": visited}
```
Comment (iamogbz) [inline src/syrupy/serializers/amber.py]: Why the change?

File: src/syrupy/serializers/amber.py [hunk 4]
```diff
@@ -146,7 +163,11 @@ def serialize_dict(
 
     @classmethod
     def serialize_iterable(
-        cls, data: "SerializableData", *, depth: int = 0, visited: Set[Any] = set()
+        cls,
+        data: "SerializableData",
+        *,
+        depth: int = 0,
+        visited: Optional[Set[Any]] = None,
```
Comment (iamogbz) [inline src/syrupy/serializers/amber.py]: Why the refactor?

File: src/syrupy/serializers/amber.py (hunk 1 above)
Comment (noahnu) [reply src/syrupy/serializers/amber.py]: no need to create a new set if the set is empty

File: src/syrupy/serializers/amber.py (hunk 3 above)
Comment (noahnu) [reply src/syrupy/serializers/amber.py]: `{}` is preferred over `dict()`. It's slightly faster. Also more lint rules ensures we're consistent (two developers write same code)

File: src/syrupy/serializers/amber.py (hunk 4 above)
Comment (noahnu) [reply src/syrupy/serializers/amber.py]: Previously we set the default arg to set(). This does not work like you'd expect. It doesn't create a "set" every function invocation, but rather re-uses a single set
